src/Grpc/JsonTranscoding/src/Microsoft.AspNetCore.Grpc.JsonTranscoding/Internal/JsonTranscodingRouteAdapter.cs

```diff
@@ -0,0 +1,196 @@
+                        var segmentType = GetSegmentType(segment);
+                        switch (segmentType)
+                        {
+                            case SegmentType.Literal:
+                                routeValueFormatTemplateParts.Add(segment);
+                                break;
+                            case SegmentType.Any:
+                                {
+                                    var parameterName = $"__Complex_{fullPath}_{i}";
+                                    tempSegments[i] = $"{{{parameterName}}}";
+
+                                    routeValueFormatTemplateParts.Add($"{{{variableParts.Count}}}");
+                                    variableParts.Add(parameterName);
+                                    break;
+                                }
+                            case SegmentType.CatchAll:
+                                {
+                                    var parameterName = $"__Complex_{fullPath}_{i}";
+                                    var suffix = string.Join("/", tempSegments.Skip(i + 1));
+                                    catchAllSuffix = string.Join("/", tempSegments.Skip(i + segmentCount - 1));
+
+                                    // It's possible to have multiple routes with catch-all parameters that have different suffixes.
+                                    // For example:
+                                    // - /{name=v1/**/b}/one
+                                    // - /{name=v1/**/b}/two
+                                    // The suffix is added as a route constraint to avoid matching multiple routes to a request.
+                                    var constraint = suffix.Length > 0 ? $":regex({suffix}$)" : string.Empty;
+                                    tempSegments[i] = $"{{**{parameterName}{constraint}}}";
+
+                                    routeValueFormatTemplateParts.Add($"{{{variableParts.Count}}}");
+                                    variableParts.Add(parameterName);
+                                    haveCatchAll = true;
+
+                                    // Remove remaining segments. They have been added in the route constraint.
+                                    while (i < tempSegments.Count - 1)
+                                    {
+                                        tempSegments.RemoveAt(tempSegments.Count - 1);
+                                    }
+                                    break;
+                                }
+                        }
+                        i++;
+                    }
+
+                    var routeValueFormatTemplate = string.Join("/", routeValueFormatTemplateParts);
+
+                    // Add an action to reconstruct the multiple segment parameter from ASP.NET Core
+                    // request route values. This should be called when the request is received.
+                    rewriteActions.Add(context =>
```

Comment:
> The object[] is completely temporary, so a good candidate for using a pool.

It's never going to be larger than 32 items. And usually much smaller than tat. Pooling would cost more time than it saves.

Reply:
> The catchAllSuffix part could be able to be rewritten to not need the Substring since you know about it before creating the rewrite action.

It's required here because the suffix comes from the catch-all route value, not the template. It has to happen per request.

Reply:
Oh I see, it comes from the route values, so technically you could substring the last route value instead of the final result which would result in smaller allocations